### PR TITLE
improvement: move ClasspathSearch and PackageIndex to mtags-shared

### PR DESCRIPTION
**Pull request by rochala**
This PR moves `ClasspathSearch` and `PackageIndex` into mtags-shared, so they can be used for testing purposes in dotty.

### CODE REVIEW

File: tests/mtest/src/main/scala/tests/TestingWorkspaceSearch.scala [hunk 1]
```diff
@@ -9,6 +9,7 @@ import scala.meta.inputs.Input
 import scala.meta.internal.metals.SemanticdbDefinition
 import scala.meta.internal.metals.WorkspaceSymbolInformation
 import scala.meta.internal.metals.WorkspaceSymbolQuery
+import scala.meta.internal.mtags.ScalametaCommonEnrichments.XtensionWorkspaceSymbolQuery
```
Comment (tgodzik) [inline tests/mtest/src/main/scala/tests/TestingWorkspaceSearch.scala:12]: Why is this needed?

File: build.sbt [hunk 2]
```diff
@@ -255,6 +255,7 @@ lazy val mtagsShared = project
     Compile / packageSrc / publishArtifact := true,
     libraryDependencies ++= List(
       "org.lz4" % "lz4-java" % "1.8.0",
+      "com.google.protobuf" % "protobuf-java" % "3.22.3",
```
Comment (tgodzik) [inline build.sbt:258]: Why is it needed?

File: mtags/src/main/scala/scala/meta/internal/mtags/SymbolIndexBucket.scala [hunk 3]
```diff
@@ -36,7 +37,7 @@ final case class SymbolLocation(
 class SymbolIndexBucket(
     toplevels: TrieMap[String, Set[AbsolutePath]],
     definitions: TrieMap[String, Set[SymbolLocation]],
-    sourceJars: ClasspathLoader,
+    sourceJars: OpenClassLoader,
```
Comment (tgodzik) [inline mtags/src/main/scala/scala/meta/internal/mtags/SymbolIndexBucket.scala:40]: Could we remove `ClasspathLoader` now actually? `ClasspathLoader` is really just used in tests and I think we could just read the classpath java property instead int hose cases.

File: mtags-shared/src/main/scala/scala/meta/internal/metals/PackageIndex.scala [hunk 4]
```diff
@@ -88,7 +84,56 @@ class PackageIndex() {
...
+   * {{{
+   *   basename("/a/b/") == "b"
+   *   basename("/a/b") == "b"
+   *   basename("/") == ""
+   * }}}
+   *
+   * @param abspath
+   *   a string path that matches the syntax of ZipFile entries.
+   */
+  def basename(abspath: String): String = {
```
Comment (tgodzik) [inline mtags-shared/src/main/scala/scala/meta/internal/metals/PackageIndex.scala:124]: Are these methods copied over from PathIO ?

**Review by tgodzik — CHANGES_REQUESTED** [review]
Great work! I left a couple of questions. If you don't have time I can follow up removing ClasspathLoader

File: tests/mtest/src/main/scala/tests/TestingWorkspaceSearch.scala (hunk 1 above)
Comment (rochala) [reply tests/mtest/src/main/scala/tests/TestingWorkspaceSearch.scala:12]: It provides extension methods which allows comparison against semanticdb SymbolInformation. I had to move it out of the `WorkspaceSymbol` class as it required semanticdb definitions which are not available in `mtags-shared`.

File: build.sbt (hunk 2 above)
Comment (rochala) [reply build.sbt:258]: It is required for `Compression` object which is used in `CompressedPackageIndex`. While I could have created e.g. an uncompressed package index to use in dotty, it just was more convenient to use the same one in both places, especially considering that `protobuf-java` is already a dependency in the compiler, and they guarantee binary compatibility between major releases.

File: mtags/src/main/scala/scala/meta/internal/mtags/SymbolIndexBucket.scala (hunk 3 above)
Comment (rochala) [reply mtags/src/main/scala/scala/meta/internal/mtags/SymbolIndexBucket.scala:40]: I'll try to remove it.

File: mtags-shared/src/main/scala/scala/meta/internal/metals/PackageIndex.scala (hunk 4 above)
Comment (rochala) [reply mtags-shared/src/main/scala/scala/meta/internal/metals/PackageIndex.scala:124]: Yes. I noticed that after introducing `mtags-shared` I had to replace all `AbsolutePath` and `RelativePath` usages to NIO. It led to a situation like this, where some code was handled in scalameta parser and forced me to copy it. It also introduced many inconsistencies in the code. Maybe it is worth addressing the whole issue by doing some refactor / unification of those paths in the code ?

File: mtags/src/main/scala/scala/meta/internal/mtags/SymbolIndexBucket.scala (hunk 3 above)
Comment (tgodzik) [reply mtags/src/main/scala/scala/meta/internal/mtags/SymbolIndexBucket.scala:40]: Thanks!

File: mtags-shared/src/main/scala/scala/meta/internal/metals/PackageIndex.scala (hunk 4 above)
Comment (tgodzik) [reply mtags-shared/src/main/scala/scala/meta/internal/metals/PackageIndex.scala:123]: This is fine, no worries.

**Comment by rochala** [comment]
Is the windows CI flaky ? I'm not sure if I made it that way or was it the case before this PR.

**Review by tgodzik — APPROVED** [review]
LGTM!
